lib/pymedphys/_mosaiq/sessions.py

```diff
@@ -0,0 +1,186 @@
+    datetime.datetime(2019, 12, 19, 2, 0)),
+    (2,
+    datetime.datetime(2019, 12, 19, 5, 0),
+    datetime.datetime(2019, 12, 19, 7, 0)),
+    (3,
+    datetime.datetime(2019, 12, 19, 10, 0),
+    datetime.datetime(2019, 12, 19, 12, 0))]
+    """
+    # turn the datetimes in to timestamps (seconds in the epoch)
+    timestamps = [[tx_datetime.timestamp()] for tx_datetime in tx_datetimes]
+
+    # set up the cluster algorithm
+    cluster_algo = sklearn.cluster.AgglomerativeClustering(
+        n_clusters=None, distance_threshold=interval.seconds, linkage="single"
+    )
+
+    # and fit the timestamps to clusters
+    labels = cluster_algo.fit_predict(timestamps)
+
+    # the resulting labels are arbitrary, so iterate and collect
+    #   all of the datetimes for each label
+    current_label, current_session_number = labels[0], 1
+    start_session = datetime.fromtimestamp(timestamps[0][0])
+    end_session = start_session
+
+    # now iterate through the labels and timestamps,
+    #   extracting each session
+    for label, timestamp in zip(labels, timestamps):
+        if label != current_label:
+            # we have a new label, so yield the current session
+            yield (current_session_number, start_session, end_session)
+
+            # and update session stats for the next one
+            current_session_number += 1
+            current_label = label
+            start_session = datetime.fromtimestamp(timestamp[0])
+            end_session = start_session
+        else:
+            # if the same label, then just update the end_session
+            #   for the new timestamp
+            end_session = datetime.fromtimestamp(timestamp[0])
+
+    # yield the final session
+    yield (current_session_number, start_session, end_session)
+
+
+def sessions_for_site(connection, sit_set_id):
+    """Determines the sessions for the given site (by SIT_SET_ID)
+
+    uses cluster_sessions after querying for the Dose_Hst.Tx_DtTm
+    that are associated to the site
+
+    Parameters
+    ----------
+    connection : pymssql connection opened with pymedphys.mosaiq.connect
+
+    sit_set_id : int
+        the SIT_SET_ID for the site of interest
+
+    Returns
+    -------
+    generated sequence of session tuples
+        same format as returned by cluster_sessions
+    """
+    result = mosaiq.execute(
+        connection,
+        """
+        SELECT
+            Tx_DtTm
+        FROM Dose_Hst
+        INNER JOIN
+            Site ON Site.SIT_ID = Dose_Hst.SIT_ID
+        WHERE
+            Site.SIT_SET_ID = %(sit_set_id)s
+        ORDER BY Dose_Hst.Tx_DtTm
+        """,
+        {"sit_set_id": sit_set_id},
+    )
+
+    # cluster_sessions expects a sorted list, so extract
+    #   the Tx_DtTm value from each row
+    dose_hst_datetimes = [row[0] for row in result]
+    return cluster_sessions(dose_hst_datetimes)
+
+
+def session_offsets_for_site(connection, sit_set_id, interval=timedelta(hours=1)):
```

Comment:
Given we now have the capacity, can we include a test that flexes this and `sessions_for_site` within this PR?